# Day 3 – SQL via Python: NYC School Data Exploration

## Step 1: Import Libraries

In [123]:
import pandas as pd
import psycopg2

## Step 2: Connect to the Database

In [124]:
# DB connection setup using hardcoded credentials (for onboarding only)
conn = psycopg2.connect(
    dbname="neondb",
    user="neondb_owner",
    password="npg_CeS9fJg2azZD",
    host="ep-falling-glitter-a5m0j5gk-pooler.us-east-2.aws.neon.tech",
    port="5432",
    sslmode="require"
)
cur = conn.cursor()

## Step 3: Run a Test Query

In [125]:
query = "SELECT * FROM nyc_schools.high_school_directory LIMIT 5;"
df = pd.read_sql(query, conn)
df.head()

/var/folders/t0/f0dxth6149d03d5n4n024r6h0000gn/T/ipykernel_33508/1731849884.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query, conn)


dbn                                        school_name    borough  \
0  27Q260          Frederick Douglass Academy VI High School     Queens   
1  21K559        Life Academy High School for Film and Music   Brooklyn   
2  16K393     Frederick Douglass Academy IV Secondary School   Brooklyn   
3  08X305                               Pablo Neruda Academy      Bronx   
4  03M485  Fiorello H. LaGuardia High School of Music & A...  Manhattan   

  building_code  phone_number    fax_number  grade_span_min  grade_span_max  \
0          Q465  718-471-2154  718-471-2890             9.0              12   
1          K400  718-333-7750  718-333-7775             9.0              12   
2          K026  718-574-2820  718-574-2821             9.0              12   
3          X450  718-824-1682  718-824-1663             9.0              12   
4          M485  212-496-0700  212-724-5748             9.0              12   

  expgrade_span_min expgrade_span_max  ... number_programs  \
0              None              None  ...               1   
1              None              None  ...               1   
2              None              None  ...               1   
3              None              None  ...               1   
4              None              None  ...               6   

                                          Location 1 Community Board  \
0  {'latitude': '40.601989336', 'longitude': '-73...              14   
1  {'latitude': '40.593593811', 'longitude': '-73...              13   
2  {'latitude': '40.692133704', 'longitude': '-73...               3   
3  {'latitude': '40.822303765', 'longitude': '-73...               9   
4  {'latitude': '40.773670507', 'longitude': '-73...               7   

  Council District Census Tract Zip Codes Community Districts  \
0               31       100802     20529                  51   
1               47          306     17616                  21   
2               36          291     18181                  69   
3               18           16     11611                  58   
4                6          151     12420                  20   

  Borough Boundaries City Council Districts Police Precincts  
0                  3                     47               59  
1                  2                     45               35  
2                  2                     49               52  
3                  5                     31               26  
4                  4                     19               12  

[5 rows x 105 columns]

In [126]:
df.columns


Index(['dbn', 'school_name', 'borough', 'building_code', 'phone_number',
       'fax_number', 'grade_span_min', 'grade_span_max', 'expgrade_span_min',
       'expgrade_span_max',
       ...
       'number_programs', 'Location 1', 'Community Board', 'Council District',
       'Census Tract', 'Zip Codes', 'Community Districts',
       'Borough Boundaries', 'City Council Districts', 'Police Precincts'],
      dtype='object', length=105)

In [127]:
query = "SELECT * FROM nyc_schools.school_demographics s WHERE s.dbn='27Q260' ;"
df = pd.read_sql(query, conn)
df.columns

/var/folders/t0/f0dxth6149d03d5n4n024r6h0000gn/T/ipykernel_33508/266311014.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query, conn)


Index(['dbn', 'Name', 'schoolyear', 'fl_percent', 'frl_percent',
       'total_enrollment', 'prek', 'k', 'grade1', 'grade2', 'grade3', 'grade4',
       'grade5', 'grade6', 'grade7', 'grade8', 'grade9', 'grade10', 'grade11',
       'grade12', 'ell_num', 'ell_percent', 'sped_num', 'sped_percent',
       'ctt_num', 'selfcontained_num', 'asian_num', 'asian_per', 'black_num',
       'black_per', 'hispanic_num', 'hispanic_per', 'white_num', 'white_per',
       'male_num', 'male_per', 'female_num', 'female_per'],
      dtype='object')

In [128]:
query = "SELECT s.ell_percent FROM nyc_schools.school_demographics s WHERE s.dbn='27Q260' ;"
df = pd.read_sql(query, conn)
df.head()

/var/folders/t0/f0dxth6149d03d5n4n024r6h0000gn/T/ipykernel_33508/3198710681.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query, conn)


Empty DataFrame
Columns: [ell_percent]
Index: []

In [129]:
df.columns

Index(['ell_percent'], dtype='object')

## Question 1. School Distribution: How many schools are there in each borough?

In [130]:
query = """
SELECT 
	hsd.borough,
	COUNT (distinct hsd.dbn) as number_of_schools
FROM nyc_schools.high_school_directory hsd
group by hsd.borough
order by number_of_schools desc

"""

df = pd.read_sql(query, conn)
df

/var/folders/t0/f0dxth6149d03d5n4n024r6h0000gn/T/ipykernel_33508/598962735.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query, conn)


borough  number_of_schools
0       Brooklyn                121
1          Bronx                118
2      Manhattan                106
3         Queens                 80
4  Staten Island                 10

## Question 2. Language Learners: What is the average % of English Language Learners (ELL) per borough?


In [131]:
query = """
select
	AVG(sd.ell_percent ) as average_percent_ELL,
	hsd.borough 
from nyc_schools.school_demographics sd 
right join nyc_schools.high_school_directory hsd
join nyc_schools.school_safety_report ssr on ssr.dbn = hsd.dbn  
on sd.dbn  = hsd.dbn 
group by hsd.borough
order by average_percent_ELL desc

"""

df = pd.read_sql(query, conn)
df

/var/folders/t0/f0dxth6149d03d5n4n024r6h0000gn/T/ipykernel_33508/3734670517.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query, conn)


average_percent_ell        borough
0                  NaN       Brooklyn
1                  NaN         Queens
2                  NaN  Staten Island
3                  NaN          Bronx
4               7.5725      Manhattan

## Question 3. School supporting special needs: Using the data from the school demographics and high school directory, write a query to find the top 3 schools in each borough with the highest percentage of special education students (sped_percent)

In [132]:
query = """
WITH Ranked_Schools AS (
  SELECT
    h.borough,
    h.school_name,
    s.sped_percent,
    ROW_NUMBER() OVER(PARTITION BY h.borough ORDER BY s.sped_percent DESC) as rank_num
  FROM
    nyc_schools.school_demographics s
    RIGHT JOIN nyc_schools.high_school_directory h ON s.dbn = h.dbn
  WHERE s.sped_percent IS NOT NULL 
)
SELECT
  borough,
  school_name,
  sped_percent
FROM
  Ranked_Schools
WHERE
  rank_num <= 3
ORDER BY
  borough,
  sped_percent DESC;

"""

df = pd.read_sql(query, conn)
df

/var/folders/t0/f0dxth6149d03d5n4n024r6h0000gn/T/ipykernel_33508/1366178868.py:27: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql(query, conn)


borough                 school_name  sped_percent
0  Manhattan  East Side Community School          28.8
1  Manhattan  East Side Community School          27.7
2  Manhattan  East Side Community School          26.7

# SQL AnalysisInsights

School Distribution Analysis

  - Brooklyn leads with 121 high schools, followed closely by the Bronx (118 schools)
  - Manhattan has 106 schools, Queens 80 schools
  - Staten Island significantly smaller with only 10 schools

  English Language Learners (ELL)

  - Data quality issue identified: Only Manhattan returned ELL data (7.57% average)
  - Other boroughs show NULL values, indicating incomplete demographic data coverage
  - Suggests potential data integration problems between school directory and demographics tables

  Special Education Services

  - Limited results: Only Manhattan schools appeared in special education analysis
  - East Side Community School shows highest special education percentages (26.7%-28.8%)
  - Data coverage gap: Missing special education data for other boroughs indicates incomplete demographic records

  Key Data Quality Findings

  - Join issues: Demographics table appears to have limited coverage compared to school directory
  - Missing records: Many schools lack demographic data, affecting borough-level analysis
  - Query optimization needed: Complex joins returning incomplete results across boroughs

  The analysis reveals significant data completeness challenges that would require data cleaning before drawing comprehensive insights about
   NYC school demographics.